zerver/lib/subscription_info.py

```diff
@@ -811,21 +814,60 @@ def get_stream_id(sub_dict: RawSubscriptionDict) -> int:
             subscribed_stream_ids,
         )
 
-        # Eventually "partial subscribers" will return (at minimum):
-        # (1) all subscriptions for recently active users (for the buddy list)
-        # (2) subscriptions for all bots
-        #
-        # For now, we're only doing (2).
-        send_partial_subscribers = include_subscribers == "partial"
+        # `partial_subscribers` returns:
+        #  * Any subscriber that would show up in `buddy_data.get_filtered_user_id_list`,
+        #    which determines who is shown in the buddy list. This ensures we show them
+        #    in the correct section. We should be careful to keep this code in sync with
+        #    that buddy list logic.
+        #  * Subscriptions for all bots, so that we can calculate human subscriber
+        #    counts on the client.
         partial_subscriber_map: dict[int, list[int]] = dict()
+
+        # This constant should stay in sync with prescence.ts
+        # (send it from the server?)
+        BIG_REALM_COUNT = 250
+        active_users = list(user_profile.realm.get_active_users())
+        send_partial_subscribers = (
+            include_subscribers == "partial" and len(active_users) >= BIG_REALM_COUNT
+        )
+
+        def get_recently_active_user_ids() -> set[int]:
+            recent = timezone_now() - timedelta(seconds=settings.OFFLINE_THRESHOLD_SECS)
+            rows = UserPresence.objects.filter(
+                last_connected_time__gte=recent,
+            ).values("user_profile_id")
+            online_user_ids = {row["user_profile_id"] for row in rows}
+            return set(online_user_ids)
```

Comment:
So rather than doing a query here unconditionally, I think we ideally want to have the `events.py` code pass this set in; we have likely already fetched it if `want("presence")` was true. Hmm, but I was also thinking we probably just want to use `UserProfile.long_term_idle` as the feature; that has advantages, including not requiring a separate feature and not being sensitive to "invisible mode".